### PR TITLE
Add  flag for specifying number of subprocesses to use to run checks in parallel

### PR DESCRIPTION
**Pull request by mschwager**
Adding Python support for #594.

### CODE REVIEW

**Review by ievans — APPROVED** [review]
Yes! This is perfect, esp defaulting to ncpus

File: semgrep/semgrep/cli.py [hunk 1]
```diff
@@ -81,6 +82,14 @@ def cli() -> None:
         action="store_true",
     )
     config.add_argument("--precommit", help=argparse.SUPPRESS, action="store_true")
+    config.add_argument(
+        "-j",
+        "--jobs",
```
Comment (brendongo) [inline semgrep/semgrep/cli.py:92]: Is `--cpus` more clear?

**Comment by brendongo** [comment]
Closes #611

File: semgrep/semgrep/cli.py (hunk 1 above)
Comment (mschwager) [reply semgrep/semgrep/cli.py:92]: I actually see `-j/--jobs` more commonly used, e.g. `make` and `flake8`

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,11 @@
 
 This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+
+### Added
+- The `-j/--jobs` flag for specifying number of subprocesses to use to run checks in parallel.
```
Comment (aryx) [inline CHANGELOG.md:8]: thx!
I wanted to add it but was waiting for it being supported in the python wrapper.

File: semgrep-core/bin/main_semgrep_core.ml [hunk 3]
```diff
@@ -109,7 +109,7 @@ let set_gc () =
   ()
 
 let map f xs =
-  if !ncores = 1
+  if !ncores <= 1
```
Comment (aryx) [inline semgrep-core/bin/main_semgrep_core.ml:162]: oops

**Review by aryx — APPROVED** [review]
Cool!

**Review by ievans — APPROVED** [review]

**Comment by mschwager** [comment]
Okay, I'm officially out of ideas here, but I've learned a few things and see some opportunity for improvement to make debugging easier in the future. I created #623 to track the division-by-zero error itself. It seems this is definitely happening somewhere other than `ncores` and needs to be fixed. Other than that:

1. It'd be very helpful to always get full OCaml exception tracebacks in CI. We may even want to always enable this even in release builds so we can get more information if end users get an exception. It seems the exception is swallowed up in the following code:

```ocaml
let format_output_exception e : string = 
  (* if (ouptut_as_json) then *)
  let msg = match e with 
    ...
    | exn -> 
      J.Object [  "error", J.String "unknown exception"; "message", J.String (Common.exn_to_s exn); ]
  in 
    Json_io.string_of_json msg
```

I tried adding `OCAMLRUNPARAM=b` at every step along the way (e.g. the Github Action layer, the bash testing layer, and the `subprocess` `semgrep-core` layer) to no avail. Without seeing *where* the division-by-zero is happening it's really difficult to fix the issue. We'll face similar issues with exceptions in the future.

2. The testing infrastructure is really difficult to grok. We have a comprehensive test suite set up to prevent regressions, but when something goes wrong it's difficult to track down. For example:

We currently have `12` Github Actions:

```
$ ls .github/workflows/ | wc -l
12
```

From there, to go from the failing action to the exercised code takes the following steps:

* The `.github/workflows/build.yml` action runs the tests
* The tests are run via a Makefile at `semgrep/Makefile`
* The Makefile runs a bash script at `semgrep/tests/run-lint-tests.sh`
* The bash script runs semgrep in test mode with `--test` at `semgrep/semgrep/cli.py`
* Semgrep runs `semgrep/semgrep/test.py` after the `--test` flag is passed
* The `test.py` script reworks much of the existing CLI flag handling
* The `test.py` script calls the main semgrep functionality in `semgrep/semgrep/semgrep_main.py` with modified CLI flags
* The main semgrep function eventually calls the semgrep-core binary from `semgrep/semgrep/core_runner.py`
* The semgrep-core binary is called - debugging OCaml issues can now begin

There is _a lot_ of complexity here, lots of which could likely be avoided. Much would be improved from #626. This would hopefully remove the need for: the test runner bash script, the `--test` flag, the `test.py` file, and all their associated functionality. Simplifying and removing unnecessary steps here would make our CI/testing/debuggability lives much easier.

There may also be some worthwhile cleanup to be done with all the Github actions. Again, simplifying, consolidating, and minimizing functionality here would help us out greatly when things go wrong.

This PR was originally meant to add a simple CLI flag to semgrep, but became a deep-dive into the codebase. Learning about all this functionality and code was definitely beneficial, but I also see some room for improvement. Tackling suggestion 1 and 2 above would go a long way to improve our testing and debuggability story.

**Comment by mschwager** [comment]
Created #634 to track suggestion 1 above. #626 seems fine for suggestion 2 for now.

**Comment by brendongo** [comment]
@aryx when running `semgrep-core` with -j 2 I get `{ "error": "unknown exception", "message": "Division_by_zero" }`

**Comment by brendongo** [comment]
@aryx I was able to reproduce a division by error in semgrep-core the following way:

Starting in repo root
`cd semgrep`
`echo blah.yaml`
>
```
rules:
- id: 0.tests.python.assert-eqeq-is-ok
  pattern: "$X == $X"
  severity: ERROR
  languages:
  - python
  message: <internalonly>
```
`semgrep-core -lang python -rules_file blah.yaml  -j 2 tests/lint`

> { "error": "unknown exception", "message": "Division_by_zero" }


Interestingly, `semgrep-core -lang python -rules_file blah.yaml  -j 2 tests/` does not get the division by zero error

**Comment by ulziibay** [comment]
And the trace comes from:
```
There was an error: Division_by_zeroRaised at file "hashtbl.ml", line 194, characters 19-28
Called from file "netexn.ml", line 78, characters 18-44
{ "error": "unknown exception", "message": "Division_by_zero" }
```

File: semgrep/semgrep/semgrep_main.py [hunk 4]
```diff
@@ -251,7 +251,7 @@ def main(args: argparse.Namespace) -> Dict[str, Any]:
 
     # actually invoke semgrep
     rule_matches_by_rule, semgrep_errors = CoreRunner(
-        args.dangerously_allow_arbitrary_code_execution_from_rules,
+        args.dangerously_allow_arbitrary_code_execution_from_rules, args.jobs
```
Comment (DrewDennison) [inline semgrep/semgrep/semgrep_main.py]:
```suggestion
        allow_exec=args.dangerously_allow_arbitrary_code_execution_from_rules, jobs=args.jobs
```

**Review by aryx — APPROVED** [review]

**Comment by DrewDennison** [comment]
@mschwager fixed the div_by_zero error with `| _ when n <= !ncores -> 1` (it came about because `1 / 2 = 0` in integer division -merging this in.

**Pull request merged**